contracts/AMMStrategy.sol

```diff
@@ -0,0 +1,181 @@
+                    _permit2Approve(tokens[i], spenders[j], amount);
+                } else {
+                    IERC20(tokens[i]).safeApprove(spenders[j], amount);
+                }
+            }
+        }
+    }
+
+    /// @inheritdoc IAMMStrategy
+    function withdrawLegacyTokensTo(address[] calldata tokens, address receiver) external override onlyOwner {
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            uint256 selfBalance = Asset.getBalance(tokens[i], address(this));
+            if (selfBalance > 0) {
+                Asset.transferTo(tokens[i], payable(receiver), selfBalance);
+            }
+        }
+    }
+
+    /************************************************************
+     *                   External functions                      *
+     *************************************************************/
+    /// @inheritdoc IStrategy
+    function executeStrategy(
+        address inputToken,
+        address outputToken,
+        uint256 inputAmount,
+        bytes calldata data
+    ) external payable override onlyEntryPoint {
+        Operation[] memory ops = abi.decode(data, (Operation[]));
+        require(ops.length > 0, "empty operations");
+        require(inputAmount > 0, "empty inputAmount");
+        uint256 balanceBefore = Asset.getBalance(outputToken, entryPoint);
+        // wrap eth first
+        if (Asset.isETH(inputToken)) {
+            IWETH(weth).deposit{ value: inputAmount }();
+        }
+        for (uint256 i = 0; i < ops.length; ++i) {
+            Operation memory op = ops[i];
+            require(ammMapping[op.dest], "invalid op dest");
+            bytes4 selector = _call(op.dest, op.value, op.data);
+            emit Action(op.dest, op.value, selector);
+        }
+        _transferToEntryPoint(outputToken);
+        uint256 balanceAfter = Asset.getBalance(outputToken, entryPoint);
+        emit Swapped(inputToken, inputAmount, outputToken, balanceAfter - balanceBefore);
+    }
+
+    /**
+     * @dev internal function of `approveTokens`.
+     * Used to permit.
+     */
+    function _call(
+        address _dest,
+        uint256 _value,
+        bytes memory _data
+    ) internal returns (bytes4 selector) {
+        if (_data.length >= 4) {
+            selector = bytes4(_data);
+        }
+        // withdraw needed native eth
+        if (_value > 0 && address(this).balance < _value) {
+            IWETH(weth).withdraw(_value - address(this).balance);
+        }
+        (bool success, bytes memory result) = _dest.call{ value: _value }(_data);
+        if (!success) {
+            assembly {
+                revert(add(result, 32), mload(result))
+            }
+        }
+    }
+
+    /**
+     * @dev internal function of `executeStrategy`.
+     * Allow the spender to use Permit2 for the token.
+     */
+    function _permit2Approve(
```

Comment:
We should first approve permit2 and then approve in permit2

Reply:
Q: So universal router will use plain `transferFrom` from permit2, without any permit signature?

Reply:
Yes, universal router utilize allowance transfer instead of signature transfer.

Reply:
Note: The UniversalRouter uses Permit2 to remove the need for token approvals being provided directly to the UniversalRouter. The Permit2 documentation can be found [here](https://docs.uniswap.org/contracts/permit2/overview).

I think so in command like `V2_SWAP_EXACT_IN`
* `address` The recipient of the output of the trade
* `uint256` The amount of input tokens for the trade
* `uint256` The minimum amount of output tokens the user wants
* `address[]` The UniswapV2 token path to trade along
* `bool` A flag for whether the input tokens should come from the msg.sender (through Permit2) or whether the funds are already in the UniversalRouter